### ISSUE TITLE
[Bug] tile.set_validshape on a cube matmul operand misclassifies the scope as MIXED, splitting it into an AIC/AIV pair

### ISSUE THREAD

**Issue by lyfne123** [hw-native-sys/pypto#2227]
## Background

In pypto-lib, `models/qwen3/14b/decode_fwd.py` on a2a3 has a **performance** problem: wrapping the matmul A operands in `pl.tensor.set_validshape(...)` — to express that only `batch` of the 16 padded rows are live — makes the decode layer ~89% more expensive (sum of per-task durations 60.3ms -> 113.8ms at batch=1, makespan 2982us -> 3441us, scheduled aicore tasks 994 -> 2494).

Root cause is not the valid-shape bookkeeping itself: it is that `set_validshape` flips the enclosing InCore scope from pure-CUBE to MIXED, after which `ExpandMixedKernel` splits every matmul function into an AIC/AIV pair.

Reproduce with (no device needed, compile-only):

    python repro_validshape_mixed.py -p a2a3            # -> FunctionType.AIC
    python repro_validshape_mixed.py -p a2a3 --valid    # -> Group{AIC, AIV}

The repro case is **not committed** — save the file below to run it.

<details>
<summary>Reproduction case — <code>repro_validshape_mixed.py</code></summary>

```python
# Minimal repro: tile.set_validshape on a cube matmul operand flips a pure-AIC
# InCore scope to MIXED, and ExpandMixedKernel then splits it into an AIC/AIV pair.
#
#   python repro_validshape_mixed.py            -> baseline, expects FunctionType.AIC
#   python repro_validshape_mixed.py --valid    -> with set_validshape, gets Group{AIC,AIV}
#
# The scope contains nothing but a matmul and its store; the ONLY difference
# between the two runs is the set_validshape wrapper on the A operand.
import argparse

import pypto.language as pl
import torch
from pypto.backend import BackendType, set_backend_type

M, K, N = 16, 256, 256
VALID_ROWS = 8

parser = argparse.ArgumentParser()
parser.add_argument("-p", "--platform", default="a2a3")
parser.add_argument("-d", "--device", type=int, default=0)
parser.add_argument("--valid", action="store_true",
                    help="wrap the matmul A operand in set_validshape")
args = parser.parse_args()

USE_VALID = args.valid
set_backend_type(BackendType.Ascend910B if args.platform.startswith("a2a3") else BackendType.Ascend950)


@pl.jit
def probe_mm(
    a: pl.Tensor[[M, K], pl.BF16],
    b: pl.Tensor[[K, N], pl.BF16],
    out: pl.Out[pl.Tensor[[M, N], pl.FP32]],
):
    with pl.at(level=pl.Level.CORE_GROUP, name_hint="probe_mm"):
        if USE_VALID:
            # The single line under test. VALID_ROWS is a compile-time constant
            # here; a runtime pl.tensor.dim(...) value behaves identically.
            acc = pl.matmul(
                pl.tensor.set_validshape(a[:, 0:K], VALID_ROWS, K),
                b[0:K, 0:N],
                out_dtype=pl.FP32,
            )
        else:
            acc = pl.matmul(a[:, 0:K], b[0:K, 0:N], out_dtype=pl.FP32)
        res = pl.assemble(out, acc, [0, 0])
    return res


post = probe_mm.compile_for_test(
    torch.zeros([M, K], dtype=torch.bfloat16),
    torch.zeros([K, N], dtype=torch.bfloat16),
    torch.zeros([M, N], dtype=torch.float32),
)
print(f"set_validshape = {USE_VALID}")
for fn in post.functions.values():
    if "probe_mm" in fn.name:
        print(f"  {fn.name}: {fn.func_type}")
```

</details>

Reproduction environment:

| Component | Version |
|---|---|
| pypto-lib | `b1233df` (branch: `feat/dynamic-public-batch-1-16-for-qwen3-14b-decode`) |
| pypto | `869d5b6` (branch: `main`) — also reproduced on `50033744` |
| simpler | `9922afdb` (detached) |
| ptoas | `0.54` |
| pto-isa | `83d01313d9bf` (from `runtime/pto_isa.pin`) |
| CANN | not detected |

Diagnosis: **pypto** — `CoreAffinity` classification / tile memory-space inference.

### Component

Backend

### Description

`core_affinity.cpp:30` decides engine affinity purely from the tile's memory space:

```cpp
bool IsCubeMemorySpace(MemorySpace ms) { return ms != MemorySpace::DDR && ms != MemorySpace::Vec; }
```

`tile.set_validshape` has no entry in `infer_tile_memory_space_pass.cpp`, so its result defaults to `Mem.Vec` -> VECTOR affinity. Combining that one statement with the matmul's CUBE affinity yields `CoreAffinity::MIXED`, and `expand_mixed_kernel_pass.cpp:1871` then declines the pure-AIC conversion:

```cpp
bool is_mixed = (combined == CoreAffinity::MIXED);
if (!is_mixed) {
    FunctionType new_type = (combined == CoreAffinity::CUBE) ? FunctionType::AIC : FunctionType::AIV;
```

The A operand here is a GM slice being fed to the cube; `set_validshape` only needs to bound the MTE2 row count and never has to land in UB, so the Vec default looks unintended rather than inherent.

There is already precedent for exactly this fix shape — `infer_tile_memory_space_pass.cpp:265-268` patches the same failure mode for `tile.create`:

> a `tile.create` conservatively defaults to `Mem.Vec` but the loop body writes a different space (e.g. Acc from matmul_acc). Without this override the final `tile.store` reads a Vec tile and **ExpandMixedKernel misclassifies the kernel as mixed, producing broken AIC/AIV IR**.

### Steps to Reproduce

1. Save the inlined `repro_validshape_mixed.py` above anywhere on the path.
2. `python repro_validshape_mixed.py -p a2a3` — prints `FunctionType.AIC`.
3. `python repro_validshape_mixed.py -p a2a3 --valid` — prints `Group` + `_aic` + `_aiv`.

The scope contains nothing but a matmul and its store; the only difference between the two runs is the `set_validshape` wrapper on the A operand.

### Expected Behavior

A scope containing only a matmul, its store, and a `set_validshape` on the matmul's operand stays `FunctionType.AIC`.

### Actual Behavior

```
baseline:  probe_mm_probe_mm: FunctionType.AIC

--valid:   probe_mm_probe_mm:     FunctionType.Group
           probe_mm_probe_mm_aic: FunctionType.AIC
           probe_mm_probe_mm_aiv: FunctionType.AIV
```

In decode_fwd this turns 16 AIV kernel files into 34 and scheduled tasks 994 -> 2494. The generated AIC and AIV bodies are byte-identical apart from the launcher (the AIV one adds the `get_subblockid()` bridge), consistent with the A2A3 mixed-task model noted in the emitted source: *"A2A3 mixed tasks run the same AIV kernel on two vector cores."*

### Git Commit ID

869d5b6

### NPU Kind

Ascend 910B

### Host Platform

Linux (aarch64)

### Additional Context

- Constness is irrelevant: a compile-time literal `set_validshape(a, 8, K)` splits exactly like a runtime `pl.tensor.dim(seq_lens, 0)` value.
- Impact beyond perf: it blocks using `valid_shape` as the way to express a partial batch, which is the natural mechanism for a padded decode pipeline's tail chunk.
- `models/qwen3/14b/decode_layer_a8w8.py` already wraps its matmul operands this way, but with `ACTIVE_BATCH == BATCH_PAD`, so that path has never run below full width.

Related: pypto#594, pypto#1031, pypto#1911, pypto-lib#115
